encoder/vaapiencoder_vp9.cpp

```diff
@@ -191,23 +192,56 @@ bool VaapiEncoderVP9::fill(VAEncPictureParameterBufferVP9* picParam,
             picParam->reference_frames[i] = VA_INVALID_SURFACE;
     }
     else {
-        picParam->refresh_frame_flags = 0x01; // refresh last frame
+
         picParam->pic_flags.bits.frame_type = kInterFrame;
 
         ReferenceQueue::const_iterator it = m_reference.begin();
 
-        for (uint32_t i = 0; it != m_reference.end(); ++it, i++)
+        for (uint32_t i = 0; it != m_reference.end(); ++it, i++) {
             picParam->reference_frames[i] = (*it)->getID();
+#if __ENABLE_DEBUG__
+            DEBUG("reference frame[%d] 0x%x", i, (*it)->getID());
+#endif
+        }
 
+        picParam->pic_flags.bits.frame_context_idx = 0;
         // last/golden/alt is used as reference frame. L0 forward
         picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
 
         // golden and alt are last KeyFrame
         // last is last decoded frame
-        picParam->ref_flags.bits.ref_last_idx = 0;
-        picParam->ref_flags.bits.ref_gf_idx = 1;
-        picParam->ref_flags.bits.ref_arf_idx = 2;
-        picParam->pic_flags.bits.frame_context_idx = 0;
+
+        if (getReferenceMode()) {
```

Comment:
Little complex here, is it ok use a 3 element queue to hold the index?
The init value for the queue Q is 0, 1, 2
For mode 0, you do not need change any thing 
For mode 1, you just need pop back and push it to front, ie 2, 0, 1 for first round.
Benefit for this solution is
The last,gf,arf allways at Q[0], Q[1], Q[2]. And the picParam->refresh_frame_flags always be 1 << Q[0] for any mode.